gateway/src/lib.rs

```diff
@@ -787,6 +796,80 @@ impl Actor {
         Ok(())
     }
 
+    /// Submit a new cron checkpoint
+    ///
+    /// It only accepts submission at multiples of `cron_period` since `genesis_epoch`, which are
+    /// set during construction. Each checkpoint will have its number of submissions tracked. The
+    /// same address cannot submit twice. Once the number of submissions is more than or equal to 2/3
+    /// of the total number of validators, the messages will be applied.
+    ///
+    /// Each cron checkpoint will be checked against each other using blake hashing.
+    fn submit_cron(rt: &mut impl Runtime, params: CronCheckpoint) -> Result<RawBytes, ActorError> {
+        // submit cron can only be performed by signable addresses
+        rt.validate_immediate_caller_type(CALLER_TYPES_SIGNABLE.iter())?;
+
+        let msgs = rt.transaction(|st: &mut State, rt| {
+            // first we check the epoch is the correct one
+            let genesis_epoch = st.genesis_epoch;
+
+            // we process only it's multiple of cron_period since genesis_epoch
+            if (params.epoch - genesis_epoch) % st.cron_period != 0 {
+                return Err(actor_error!(illegal_argument, "epoch not allowed"));
+            }
+
+            let store = rt.store();
+            let submitter = rt.message().caller();
+
+            st.cron_submissions
+                .modify(store, |hamt| {
+                    let epoch_key = BytesKey::from(params.epoch.to_be_bytes().as_slice());
+                    let mut submission = match hamt.get(&epoch_key)? {
+                        Some(s) => s.clone(),
+                        None => CronSubmission::new(store)?,
+                    };
+
+                    let epoch = params.epoch;
+                    let reached_limit = submission.submit(store, submitter, params)?;
+
+                    if !reached_limit || st.last_cron_executed_epoch + st.cron_period != epoch {
+                        hamt.set(epoch_key, submission)?;
+                        return Ok(None);
+                    }
+
+                    st.last_cron_executed_epoch = epoch;
+
+                    let msgs = submission
+                        .load_most_submitted_checkpoint(store)?
+                        .unwrap()
+                        .top_down_msgs;
+                    hamt.set(epoch_key, submission)?;
+
+                    Ok(Some(msgs))
+                })
+                .map_err(|e| {
+                    log::error!(
+                        "encountered error processing submit cron checkpoint: {:?}",
+                        e
+                    );
+                    actor_error!(unhandled_message, e.to_string())
+                })
+        })?;
+
+        if let Some(msgs) = msgs {
+            // TODO: we might need to batch the execution so that it's atomic
```

Comment:
It will be atomic in any case, as the execution of `submit_cron` will be atomic in itself. If any of these top-down messages fail, the whole submit_cron will fail. We may have to figure out how to handle when this happens in the future, i.e. what if one of the messages in the batch of top-down messages fail.